File: UICPC/13/c.cpp
#include <bits/stdc++.h>

using namespace std;

bool inrange(int i, int j,int m, int n){
    if ((i>=0) && (i<m) && (j>=0) && (j<n)){
        return true;
    }
    else{
        return false;
    }
}


int main(){
    int m,n;
    cin>>m>>n;
    int mp[m][n];
    for (int i=0; i<m; i++){
        for (int j=0; j<n; j++){
            cin>>mp[i][j];
        }
    }
    int x,y;
    cin >>x>>y;
    x--;
    y--;
    vector <int> cars;
    vector <int> visited;
    bool flag = true;

    while (flag){
        bool pflag = true;

        if (inrange(x,y+1,m,n)){
            if (mp[x][y+1] == mp[x][y]){
                pflag = false;
                cars.push_back(mp[x][y]);
                if (inrange(x,y+2,m,n)){
                    if (mp[x][y+2]==-1){
                        for(int p = cars.size() -1 ; p >= 0 ; p--){
                            cout<<cars[p]<<' ';
                        }
                        return 0;
                    }
                    else if (mp[x][y+2]==-2){
                        cout<<"impossible";
                        return 0;
                    }
                    else {
                        if (find(visited.begin(), visited.end(), mp[x][y+2]) != visited.end()){
                            cout<<"impossible";
                            return 0;
                            }
                        else {
                            visited.push_back(mp[x][y]);
                            y+=2;
                            }
                        }
                    
                    }
                else {
                        cout<<"impossible";
                        return 0;
                    }
                }

            }

        if ((inrange(x,y-1,m,n)) && pflag){
            if (mp[x][y-1] == mp[x][y]){
                pflag = false;
                cars.push_back(mp[x][y]);
                if (inrange(x,y,m,n)){
                    if (mp[x][y-2]==-1){
                        for(int p = cars.size() -1 ; p >= 0 ; p--){
                            cout<<cars[p]<<' ';
                        }
                        return 0;
                    }
                    else if (mp[x][y-2]==-2){
                        cout<<"impossible";
                        return 0;
                    }
                    else {
                        if (find(visited.begin(), visited.end(), mp[x][y-2]) != visited.end()){
                            cout<<"impossible";
                            return 0;
                            }
                        else {
                            visited.push_back(mp[x][y-2]);
                            y-=2;
                            }
                        }
                    
                    }
                else {
                        cout<<"impossible";
                        return 0;
                    }
                }

            }

        if ((inrange(x+1,y,m,n)) && pflag){
            if (mp[x+1][y] == mp[x][y]){
                pflag = false;
                cars.push_back(mp[x][y]);
                if (inrange(x+2,y,m,n)){
                    if (mp[x+2][y]==-1){
                        for(int p = cars.size() -1 ; p >= 0 ; p--){
                            cout<<cars[p]<<' ';
                        }
                        return 0;
                    }
                    else if (mp[x+2][y]==-2){
                        cout<<"impossible";
                        return 0;
                    }
                    else {
                        if (find(visited.begin(), visited.end(), mp[x+2][y]) != visited.end()){
                            cout<<"impossible";
                            return 0;
                            }
                        else {
                            visited.push_back(mp[x+2][y]);
                            x+=2;
                            }
                        }
                    
                    }
                else {
                        cout<<"impossible";
                        return 0;
                    }
                }

        }

        if ((inrange(x-1,y,m,n)) && pflag){
            if (mp[x-1][y] == mp[x][y]){
                pflag = false;
                cars.push_back(mp[x][y]);
                if (inrange(x-2,y,m,n)){
                    if (mp[x-2][y]==-1){
                        for(int p = cars.size() -1 ; p >= 0 ; p--){
                            cout<<cars[p]<<' ';
                        }
                        return 0;
                    }
                    else if (mp[x-2][y]==-2){
                        cout<<"impossible";
                        return 0;
                    }
                    else {
                        if (find(visited.begin(), visited.end(), mp[x-2][y]) != visited.end()){
                            cout<<"impossible";
                            return 0;
                            }
                        else {
                            visited.push_back(mp[x][y]);
                            x-=2;
                            }
                        }
                    
                    }
                else {
                        cout<<"impossible";
                        return 0;
                    }
                }

            }
        if(pflag) {
            cout<<"impossible";
            return 0;
        }
        }
    }

File: UICPC/13/c.py

def inrange(i,j):
    if (i>=0) and (i<m) and (j>=0) and (j<n):
        return 1
    else:
        return 0

m,n = map(int,input().split())
dp=[]
for i in range(m):
    dp.append(list(map(int,input().split())))
x,y = map(int,input().split())
x-=1
y-=1
cars = []
visited = []
flag = True
#print(dp[0])
#print(dp[0][0])
while(flag):
    pflag = True
    if inrange(x,y+1):
        if dp[x][y+1]==dp[x][y]:
            pflag=False
            cars.append(dp[x][y])
            if inrange(x,y+2):
                if dp[x][y+2] == -1:
                    for i in range(len(cars)-1,-1,-1):
                        print(cars[i],end=' ')
                    exit()
                elif dp[x][y+2] == -2:
                    print('impossible')
                    exit()
                else:
                    if [x,y+2] in visited:
                        print("impossible")
                        exit()
                    else:
                        visited.append([x,y+2])
                        y+=2
                    
            else:
                print("impossible")
                exit()
    
    if inrange(x,y-1) and pflag:
        if dp[x][y-1]==dp[x][y]:
            pflag=False
            cars.append(dp[x][y])
            if inrange(x,y-2):
                if dp[x][y-2] == -1:
                    for i in range(len(cars)-1,-1,-1):
                        print(cars[i],end=' ')
                    exit()
                elif dp[x][y-2] == -2:
                    print('impossible')
                    exit()
                else:
                    if [x,y-2] in visited:
                        print("impossible")
                        exit()
                    else:
                        visited.append([x,y-2])
                        y-=2
            else:
                print("impossible")
                exit()
        
    if inrange(x+1,y)and pflag:
        if dp[x+1][y]==dp[x][y]:
            pflag=False
            cars.append(dp[x][y])
            if inrange(x+2,y):
                if dp[x+2][y] == -1:
                    for i in range(len(cars)-1,-1,-1):
                        print(cars[i],end=' ')
                    exit()
                elif dp[x+2][y] == -2:
                    print('impossible')
                    exit()
                else:
                    if [x+2,y] in visited:
                        print("impossible")
                        exit()
                    else:
                        visited.append([x+2,y])
                        x+=2
            else:
                print("impossible")
                exit()
            
    
    if inrange(x-1,y) and pflag:
        if dp[x-1][y]==dp[x][y]:
            pflag=False
            cars.append(dp[x][y])
            if inrange(x-2,y):
                if dp[x-2][y] == -1:
                    for i in range(len(cars)-1,-1,-1):
                        print(cars[i],end=' ')
                    exit()
                elif dp[x-2][y] == -2:
                    print('impossible')
                    exit()
                else:
                    if [x-2,y] in visited:
                        print("impossible")
                        exit()
                    else:
                        visited.append([x-2,y])
                        x-=2
            else:
                print("impossible")
                exit()
    
    if pflag:
        flag = False
        print("impossible")
        exit()